xbmc/video/videosync/VideoSyncOsx.cpp

```diff
@@ -20,135 +20,112 @@
-  setupOk = true;
-#else
-  setupOk = Cocoa_CVDisplayLinkCreate((void*)DisplayLinkCallBack, reinterpret_cast<void*>(this));
+  setupOk = InitDisplayLink();
   if (setupOk)
+  {
     g_Windowing.Register(this);
-  else
-    CLog::Log(LOGDEBUG, "CVideoSyncCocoa: Cocoa_CVDisplayLinkCreate failed");
-#endif
-
-  if (setupOk)
-    GetFps();
-
+  }
+  
   return setupOk;
 }
 
-void CVideoSyncCocoa::Run(volatile bool& stop)
+void CVideoSyncOsx::Run(volatile bool& stop)
 {
   //because cocoa has a vblank callback, we just keep sleeping until we're asked to stop the thread
   while(!stop && !m_abort)
   {
-    Sleep(1000);
+    Sleep(100);
   }
 }
 
-void CVideoSyncCocoa::Cleanup()
+void CVideoSyncOsx::Cleanup()
 {
-  CLog::Log(LOGDEBUG, "CVideoSyncCocoa: cleaning up Cocoa");
-#if defined(TARGET_DARWIN_IOS)
-  g_Windowing.DeinitDisplayLink();
-#else
-  Cocoa_CVDisplayLinkRelease();
+  CLog::Log(LOGDEBUG, "CVideoSyncOsx::%s cleaning up OSX", __FUNCTION__);
+  DeinitDisplayLink();
   g_Windowing.Unregister(this);
-#endif
 }
 
-void CVideoSyncCocoa::UpdateFPS(double fps)
+float CVideoSyncOsx::GetFps()
 {
-  int fpsInt = MathUtils::round_int(fps);
-
-  if (fpsInt != MathUtils::round_int(m_fps))
-  {
-    CLog::Log(LOGDEBUG, "CVideoSyncCocoa: Detected refreshrate: %i hertz", fpsInt);
-    m_fps = fpsInt;
-  }
+  m_fps = g_graphicsContext.GetFPS();
+  CLog::Log(LOGDEBUG, "CVideoSyncOsx::%s Detected refreshrate: %f hertz", __FUNCTION__, m_fps);
+  return m_fps;
 }
 
-void CVideoSyncCocoa::OnResetDevice()
+void CVideoSyncOsx::OnResetDevice()
 {
   m_abort = true;
 }
 
-float CVideoSyncCocoa::GetFps()
+void CVideoSyncOsx::VblankHandler(int64_t nowtime)
 {
-#if defined(TARGET_DARWIN_IOS)
-  UpdateFPS(g_Windowing.GetDisplayLinkFPS() + 0.5);
-#else
-  UpdateFPS(Cocoa_GetCVDisplayLinkRefreshPeriod());
-#endif
+  int           NrVBlanks;
+  double        VBlankTime;
+  
+  //calculate how many vblanks happened
+  VBlankTime = (double)(nowtime - m_LastVBlankTime) / (double)g_VideoReferenceClock.GetFrequency();
+  NrVBlanks = MathUtils::round_int(VBlankTime * m_fps);
+  
+  //save the timestamp of this vblank so we can calculate how many happened next time
+  m_LastVBlankTime = nowtime;
+  
+  //update the vblank timestamp, update the clock and send a signal that we got a vblank
+  UpdateClock(NrVBlanks, nowtime);
+}
 
-  return m_fps;
+// Called by the Core Video Display Link whenever it's appropriate to render a frame.
+static CVReturn DisplayLinkCallBack(CVDisplayLinkRef displayLink, const CVTimeStamp* inNow, const CVTimeStamp* inOutputTime, CVOptionFlags flagsIn, CVOptionFlags* flagsOut, void* displayLinkContext)
+{
+  // Create an autorelease pool (necessary to call into non-Obj-C code from Obj-C code)
+  void* pool = Cocoa_Create_AutoReleasePool();
+  
+  CVideoSyncOsx *VideoSyncOsx = reinterpret_cast<CVideoSyncOsx*>(displayLinkContext);
+  // TODO - is it better to use CurrentHostCounter here?!?
+  VideoSyncOsx->VblankHandler(inOutputTime->hostTime);
```

Comment:
not really. this just counts the vblanks. we assume vertical retrace at the time when UpdateClock is called. the clock is fine adjusted in a way that the discrete time is most accurate in the middle of two vblanks. to that time we sync audio.
can you log both times (and the diff between them) when playing some 24Hz video?